docs(angular): blog post for using angular with rspack (#2)
Coly010/nx: docs/blog/2025-03-19-using-angular-with-rspack.md

---
title: 'Using Angular with Rspack'
date: 2025-03-19
slug: using-angular-with-rspack
authors: [Colum Ferry]
tags: [angular, webpack, rspack, nx]

> cover_image: /blog/images/2025-03-19/angular-rspack-logo-small.png

Reviewer: we might have to create a more evolved cover image. Like one with Mike's Angular week logo and the Rspack logo on the side for example

Author: 100% - this is just a placeholder until we have that image ready

description: Learn about how and why to use Rspack with Angular thanks to Nx's efforts for supporting Rspack for Angular.
---

> Building applications with [Angular](https://angular.dev) has always been a lesser concern for most Angular developers due to the abstractions that Angular created called `builders`. The underlying implementation details were hidden from the developer who only needed to run either `ng build` or `nx build`.  

Reviewer: Minor:

Should we be more specific about "Building applications". At a first glance this looked like building applications as in creating applications with Angular, not the actual build process itself.

Maybe "configuring your build tooling" or something like that?

Author: Hmm yeah I can see the confusion. Let’s change it

> Despite this, most Angular developers knew that it was originally [webpack](https://webpack.js.org/) that was used to build their applications. This was a great solution at the time and it was even possible to extend their builds by leveraging custom webpack configurations and plugins.

Reviewer: ```suggestion
Despite this, most Angular developers knew that it was originally [Webpack](https://webpack.js.org/) that was used to build their applications. This was a great solution at the time and it was even possible to extend their builds by leveraging custom webpack configurations and plugins.
```

> Over time, as applications grew in size and complexity, it became clear that the inherit slowness with Webpack build speeds was becoming more and more of an issue for Angular developers.

Reviewer: >  build speed issue by building out a new build pipeline

😅 idk why but this reminded me of https://youtube.com/clip/UgkxI9oC4W7_NCEFIekznXVdOz2EyE6nNbEr?si=qWSBM0R3o8WShbCP

The Angular Team decided to address this build speed issue by building out a new build pipeline that leveraged [Esbuild](https://esbuild.github.io/).

> This succeeded in reducing the build times for Angular applications, however, it made one crucial mistake. It left the existing Angular applications that relied on the Webpack ecosystem behind, with either a difficult migration path or none at all.

Reviewer: I think this would be worth emphasizing more. On the enterprise articles we have a component for large quotes:

![image](https://github.com/user-attachments/assets/c042ffa3-9baa-44f0-8278-dd09c54484f3)

I think we could use the same, but without the author obviously. Alternatively we can just use a blockquote

Author: Both paragraphs or the one about the leaving webpack users behind?

Reviewer: Maybe paraphrase and summarize it. Like

> Esbuild brought much-needed performance improvements to Angular builds, but existing Webpack-based applications were left with either a difficult migration path or no clear upgrade strategy.

or ideally even shorter. but something like this

## What about Rspack?

[Rspack](https://rspack.dev) is a high performance JavaScript bundler written in Rust. It offers strong compatibility with the Webpack ecosystem, allowing for almost seamless replacement of webpack, and provides lightning fast build speeds.

Because it supports the existing Webpack ecosystem, it provides an answer to teams that maintain Angular applications using Webpack and want to migrate to a faster build pipeline.

However, it is crucial to understand that Rspack is not _completely_ compatible with the Webpack ecosystem with some slight nuances and low-level api differences that prevent certain plugins and loaders from working out of the box.
One such example is the `AngularIvyPlugin` which is a Webpack plugin that is used to support Angular's Ivy compiler. Therefore, it was not possible to simply drop in Rpsack and expect it to work with Angular applications.
Many people have tried to get Rspack working with Angular, but it has proven to be a challenge, with partial support and success being reported in the past but with a lot of limitations or actual performance degradations over the Webpack approach.

> Instead, a new approach was needed to support Rspack with Angular. A closer examination and replication of how Angular compiles and bundles the application was required.

Reviewer: This sentence is kinda weird. I get what it is hinting at, but can we be more specific. Like saying current tentatives focused on ..... while instead it was necessary to .... (which Angular Rspack does)

Author: I will have a think about how to rephrase it

## Introducing Angular Rspack

![Angular Rspack Logo](/blog/images/2025-03-19/nx-angular-rspack-logo-small.png)

Angular Rspack started in September 2024 after I spent way too long investigating and researching how exactly Angular compiles and bundles for both their Webpack support and Esbuild support.

Something that I kept coming back to was that any Rspack solution that relied too much on Angular's Webpack support had the chance of being dropped by the Angular team as they continue to build out incredible new features with Esbuild. Instead, I decided that replicating and utilizing the abstractions the Angular team provided for their Esbuild support had a much stronger chance of longevity.

We are thrilled to announce that Angular Rspack is now being maintained by the Nx team and is available for use with Nx. This means that you can now use Rspack with Angular and Nx and enjoy the benefits of both.  

> The new package is called [@nx/angular-rspack](https://www.npmjs.com/package/@nx/angular-rspack) and it is available on npm while the repository has been moved to [nrwl/angular-rspack](https://github.com/nrwl/angular-rspack).

Reviewer: Hmm...these paragraphs don't flow well. 

Probably because you start writing from your own perspective ("...after I spent way too long..") but then you unexpectedly flip from talking about what Angular Rspack is about to "We are thrilled to announce ..." :)

Maybe slightly rewrite

- always writing from ur 1st person perspective
- emphasizing this was your personal side-project
- Before you start with "We are thrilled..." mention something like "but before we dive into the technical details I am excited to announce" basically from your perspective as it moves from ur project to an Nx project
- we could mention that Nx supports already integrates with a wide variety of build tooling across the ecosystem (also filling in gaps where the Dx needs to be improved) which is why maintaining Angular Rspack is a good fit etc... something like that :)

Author: Ok yeah I like it!

We believe that a large portion of Angular applications still using Webpack will be able to migrate to Angular Rspack without issue.

There are some limitations and missing features that are currently being worked on and on the roadmap to support. They have been listed at the bottom of this article.

> ## Migrating from Angular Webpack to Angular Rspack

Reviewer: I'd love another section before this that talks about the technical details under the hood. Like how Angular Rspack integrates into the Angular build pipeline. I'm very curiuos about how that works, like the actual low-level tech details and I bet many other Angular devs are too. 

That might also be the section that outlines some of the current limitations and improvements that are being worked on.

Then I'd shuffle the sections a little like in this order

- Using Angular Rspack
- Build and Serving your Application
- Running a configuration
- Benchmarks
- (Known Limitations and Missing Features) => I'd wire this into the "how it works" section
- Migrating from Angular Webpack to Angular Rspack

> We have worked hard to make the migration process as smooth as possible.
> A new generator has been added to the `@nx/angular` package called `convert-to-rspack` which will help you migrate your Angular applications from Webpack to Rspack.

Reviewer: I'd not use "We" here. Idk it feels odd to me when reading as I read this as your blog post on the Nx blog.
So I'd just keep it more generic like

```suggestion
To make the migration process as smooth as possible, a new generator has been added to the `@nx/angular` package called `convert-to-rspack` which will help you migrate your Angular applications from Webpack to Rspack.
```

The steps are very simple:

1. Run `nx migrate latest` to update your workspace to the latest version of Nx.
2. Run `nx g @nx/angular:convert-to-rspack` to migrate your Angular application to Rspack.

> We have also created a [guide](/recipes/angular/rspack/migrate-from-webpack) that walks you through the process step-by-step.

Reviewer: ```suggestion
There is also a [guide in our documentation](/recipes/angular/rspack/migrate-from-webpack) that walks you through the process step-by-step.
```

Even if you're currently using the Angular CLI, it's as simple as first running `npx nx init` in your workspace and then running `npx nx convert-to-rspack`.

## Using Angular Rspack

> You'll notice a that your `build` and `serve` targets have been removed from your project, and are instead inferred by the `@nx/rspack/plugin`.
> You should also notice a new `rspack.config.ts` file in your project which looks something like this:

Reviewer: ```suggestion
You'll notice that after migrating to Angular Rspack your `build` and `serve` targets have been removed from your project, and are instead [inferred](/concepts/inferred-tasks) by the `@nx/rspack/plugin`.

In addition, a new `rspack.config.ts` file has been created in your project which looks something like this:
```

```ts
import { createConfig } from '@nx/angular-rspack';
export default createConfig(
  {
    options: {
      root: __dirname,
      outputPath: {
        base: '../../dist/apps/app',
      },
      index: './src/index.html',
      browser: './src/main.ts',
      polyfills: ['zone.js'],
      tsConfig: './tsconfig.app.json',
      assets: [
        './src/favicon.ico',
        './src/assets',
        {
          input: './public',
          glob: '**/*',
        },
      ],
      styles: ['./src/styles.scss'],
      scripts: [],
    },
  },
  {
    production: {
      options: {
        outputPath: {
          base: '../../dist/apps/app-prod',
        },
        index: './src/index.prod.html',
        browser: './src/main.prod.ts',
        tsConfig: './tsconfig.prod.json',
      },
    },
  }
);
```

> {% callout type="info" title="createConfig Information" %}

Reviewer: optional. For these "expanded information blocks" I often use the `type="deepdive"` which is made on purpose for  when you want to go on a little tangent and expand on something. Those are collapsed by default. 

it's optional for this one as it is super short, but you could still use it and even expand a little more about what it does potentially.

Author: I didn’t know we had a type deep dive, I like it!

Reviewer: docs/README.md is the hidden treasure of knowledge :)

The `createConfig` function is used to create an Rspack configuration object setup for Angular applications.
You can read more about it [here](/nx-api/angular-rspack/documents/create-config).
{% /callout %}

### Build and Serving your Application

> You can now run `nx build app` and `nx serve app` to build and serve your application via Rspack.

Reviewer: should we include a terminal screenshot showing the output of a serve where Rspack is shown? Just to add some images as well :)

Author: Does it need to be a screenshot, we have md support for terminal right?

Reviewer: yeah we can. let's see what looks better. screenshot pops more but I don't have strong opinions. I'd only add it if it grabs attention otherwise let's just leave it out :)

> ### Running a configuration

Reviewer: "Running" a configuration sounds weird. Maybe "Using"?

Author: Good point

To run the build with the `production` configuration:

```bash
NGRS_CONFIG=production nx build app
```

> {% callout type="info" title="NGRS_CONFIG" %}

Reviewer: I would have this as part of the normal flow without a callout necessarily.

Author: Ok, no problem!

`NGRS_CONFIG` is an environment variable that you can use to specify which configuration to use. If the environment variable is not set, the `production` configuration is used by default.
{% /callout %}

## Benchmarks

Below is a table of benchmarks for the different bundlers available, run against an application consisting of ~800 lazy loaded routes with ~10 components each - totaling ~8000 components.

Using a M3 Macbook Pro

| Build/Bundler | Prod SSR (s) | Prod (s) | Dev (s) |
| ------------- | ------------ | -------- | ------- |
| Webpack       | 348.707      | 224.226  | 234.449 |
| esbuild       | 28.509       | 24.521   | 18.719  |
| Rsbuild       | 24.690       | 20.490   | 19.675  |

> | Rspack        | 19.974       | 18.239   | 16.477  |

Reviewer: Optional: we could consider having a simple chart alongside this.

Author: Yeah, once we get the updated numbers (in progress) I can create a chart.
Do we have md support for it? I’m concerned about light vs dark mode
